### DISCUSSION TITLE
FATAL: remaining connection slots are reserved for non-replication superuser connections.

### DISCUSSION THREAD

**Discussion by ohuu** [saleor/saleor#10100] [Deployment]
I've deployed Saleor to Digital Ocean and have setup a Postgres Cluster (again in Digital Ocean). Everything works fine for a few minutes and then I hit this error.

```
FATAL: remaining connection slots are reserved for non-replication superuser connections.
```

I found this issue `https://github.com/saleor/saleor/discussions/8173` which is exactly the problem I'm facing but thought I'd open a new discussion because that one looked stale.

After looking into the problem for a bit I found that all the database connections are being used up by idle queries which never terminate. I currently have 47 connections without a connection pool (there are issues using cursors with connection pools) but I don't think the solution is to simply use a connection pool or increase the number of connections because these idle connections accumulate and will eventually use all available connections regardless of how many you provide.

The queries that get "stuck" look like this:

```
datid            | xxxx
datname          | xxxx
pid              | xxxx
leader_pid       |
usesysid         | xxxx
usename          | xxxx
application_name |
client_addr      | xx.xx.xx.xx
client_hostname  |
client_port      | xxxx
backend_start    | 2022-07-04 10:30:07.684888+00
xact_start       |
query_start      | 2022-07-04 10:30:07.737256+00
state_change     | 2022-07-04 10:30:07.739338+00
wait_event_type  | Client
wait_event       | ClientRead
state            | idle
backend_xid      |
backend_xmin     |
query_id         | -7027771156210260856
query            | SELECT "channel_channel"."id", "channel_channel"."name", "channel_channel"."is_active", "channel_channel"."slug", "channel_channel"."currency_code", "channel_channel"."default_country", EXISTS(SELECT (1) AS "a" FROM "order_order" U0 WHERE U0."channel_id" = "channel_channel"."id" LIMIT 1) AS "has_orders" FROM "channel_channel" WHERE "channel_channel"."id" IN (1)
backend_type     | client backend
```

I'm not sure why they are getting blocked. Any ideas?

**Comment by MaxTeiger** [comment]
Hello @ohuu ! 👋🏼 

Did you find a solution to this error ? I have exactly the same problem but on GCP 

Thank you 😄

**Comment by JannikZed** [comment]
We are facing the issue as well. We need to deploy 30+ saleor core docker containers and use a pgbouncer as pooler.. but however we try to tune it, we are still facing issues. Transaction mode doesn't seem to work because of the cursors problem, session mode is not working because of the connections that stay open that long.

**Comment by ohuu** [comment]
Hi @MaxTeiger & @JannikZed. I fixed it by using PGBouncer. However, PGBouncer will only work if you switch off client-side cursors. I'm not a python expert or anything and I can't quite remember exactly what I did but I think you just need to add this line `'DISABLE_SERVER_SIDE_CURSORS': True` to the DATABASES map in `settings.py`

**Comment by JannikZed** [comment]
Hi Oliver,
Good to hear, that you could make it run with disabled cursors. We were not
sure, if that might break something else somewhere.. right now, we are
using the PGBouncer with session mode and a pool of 100 connections and
increased the max connections in Postgres a little bit. Works for 20 Saleor
core containers and around 100 parallel checkouts.
Did you create the settings.py by yourself or did you find a better
solution to set the 'DISABLE_SERVER_SIDE_CURSORS'?

**Comment by timuric** [comment]
We have updated Saleor Cloud pricing, which starts from $45 a month, which can be cheaper than DO and provides hassle free updates, backups, autoscale, sandboxes, snapshot management and more. If you are not running forked version of Saleor it is worth to check out.

**Reply by MaxTeiger** [reply]
Thank you ! I think I am going to use this new option 😄
